cmd/notify-mailer/main.go

```diff
@@ -231,87 +254,138 @@ func emailsForReg(id int, dbMap dbSelector) ([]string, error) {
 
-// readRecipientsList reads a CSV filename and parses that file into a list of
-// recipient structs. It puts any columns after the first into a per-recipient
-// map from column name -> value.
-func readRecipientsList(filename string) ([]recipient, error) {
-	f, err := os.Open(filename)
+// parseRecipientList parses the contents of a recipient list file into a list
+// of `recipient` objects. All columns after `id` into a per-recipient map from
+// column name -> value.
+func parseRecipientList(recipientList *csv.Reader) ([]recipient, error) {
+	entries, err := recipientList.Read()
 	if err != nil {
 		return nil, err
 	}
-	reader := csv.NewReader(f)
-	record, err := reader.Read()
-	if err != nil {
-		return nil, err
-	}
-	if len(record) == 0 {
-		return nil, fmt.Errorf("no entries in CSV")
-	}
-	if record[0] != "id" {
-		return nil, fmt.Errorf("first field of CSV input must be an ID.")
+
+	if entries[0] != "id" {
+		return nil, fmt.Errorf(
+			"the first column of the header row must be \"id\", got: %q", entries[0])
 	}
-	var columnNames []string
-	for _, v := range record[1:] {
-		columnNames = append(columnNames, strings.TrimSpace(v))
+
+	// Collect column names used for the `Extra` field of the `recipient`
+	// object. All columns after `id` are considered `Extra`.
+	header := entries[1:]
+	var extraColNames []string
+	for _, entry := range header {
+		colName := strings.TrimSpace(entry)
+		if len(colName) == 0 {
+			return nil, fmt.Errorf(
+				"header: %q contains duplicate or trailing delimiter", header)
+		}
+		extraColNames = append(extraColNames, colName)
 	}
 
-	results := []recipient{}
+	var recipients []recipient
 	for {
-		record, err := reader.Read()
-		if err == io.EOF {
-			if len(results) == 0 {
-				return nil, fmt.Errorf("no entries after the header in CSV")
-			}
-			return results, nil
-		}
+		entry, err := recipientList.Read()
 		if err != nil {
+			if errors.Is(err, io.EOF) {
+				// Finished parsing the file.
+				break
+			}
 			return nil, err
 		}
-		if len(record) == 0 {
-			return nil, fmt.Errorf("empty line in CSV")
-		}
-		if len(record) != len(columnNames)+1 {
-			return nil, fmt.Errorf("Number of columns in CSV line didn't match header columns."+
-				" Got %d, expected %d. Line: %v", len(record), len(columnNames)+1, record)
-		}
-		id, err := strconv.Atoi(record[0])
+
+		// Attempt to coerce the value provided for `id` to a valid registration
+		// ID (int64).
+		entryID := entry[0]
+		id, err := strconv.ParseInt(entryID, 10, 64)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf(
+				"%q couldn't be parsed as a registration ID: %s", entryID, err)
 		}
-		recip := recipient{
-			id:    id,
-			Extra: make(map[string]string),
+
+		// Ensure our file doesn't contain duplicate entries for the same
+		// registration ID.
+		for _, recipient := range recipients {
+			if id == recipient.id {
+				return nil, fmt.Errorf("id: %d is duplicated", id)
+			}
```

Comment:
This is a potentially very expensive O(n^2) operation. Or rather, this operation is O(n), but it gets run n times, so the overall time is O(n^2). The general fix to such problems is to use a map for lookups.